### PR TITLE
migrate from databaseUsername to databaseAccount

### PR DESCRIPTION
**Pull request by zzzeek**
This moves placement to fully use MariaDBAccount based on the dev work being done for mariadb-operator:

https://github.com/openstack-k8s-operators/mariadb-operator/pull/184/files

Lead Jira: [OSPRH-4095](https://issues.redhat.com/browse/OSPRH-4095)

1. [x] controller calls EnsureMariaDBAccount up front to make sure MariaDBAccount is present
2. [x] error code from EnsureMariaDBAccount is handled, Conditions are amended when error is returned
3. [x] controller calls NewDatabaseForAccount instead of NewDatabase
4. [x]  GetAccountAndSecret is used to retrieve account /secret to populate template
5. [x]  GetDatabaseByName() , normally used for delete finalizers, replaced with GetDatabaseByNameAndAccount
6. [x]  CreateOrPatchAll() used to patch the Database, replacing CreateOrPatchDB / CreateOrPatchDBByName
7. [x]  controller calls DeleteUnusedMariaDBAccountFinalizers when launched pods are definitely running on a new MariaDBAccount, returns error code if present
8. [x]  PasswordSelectors that refer to database are removed
9. [x]  all databaseUser replaced with databaseAccount inside of all XYZ_types.go
10. [x] all databaseUser replaced with databaseAccount inside of all `kuttl/*.yaml`
11. [x] all default databaseUser names become databaseAccount, replacing underscores with dashes inside of all XYZ_types.go
12. [x] all default databaseUser names become databaseAccount, replacing underscores with dashes inside of all `kuttl/*.yaml`
13. [x] MariaDBAccountSuiteTests are used in controller ginkgo tests if it has them
14. [x] Use configsecrets for database URLs; remove from job hash  - already was like that
15. [x] [184](https://github.com/openstack-k8s-operators/mariadb-operator/pull/184) is merged and replaces from go.mod are removed

### CODE REVIEW

File: controllers/placementapi_controller.go
Comment (zzzeek) [inline controllers/placementapi_controller.go]: oh this is obsolete now, merge artifact

File: tests/functional/placementapi_controller_test.go
Comment (zzzeek) [inline tests/functional/placementapi_controller_test.go]: whats this, merge artifact?  fix that....

File: tests/functional/base_test.go
Comment (zzzeek) [inline tests/functional/base_test.go]: dont hardcode this, there are mistakes in the test suite because this is the same as database name.  use a variable with a distinct name

File: controllers/placementapi_controller.go
Comment (gibizer) [inline controllers/placementapi_controller.go:359]: do we need to set the requireTLS flag based on some input?

File: controllers/placementapi_controller.go
Comment (gibizer) [inline controllers/placementapi_controller.go:357]: I feel that ensuring the DBAccount logically belongs to ensureDB, so I would move this new logic into that func.

File: tests/functional/placementapi_controller_test.go
Comment (gibizer) [inline tests/functional/placementapi_controller_test.go]: As golang does not have keyword args these functions does not have a name here to signify their purpose and 
 their requirements. Would it be more readable to pass a struct with func fields to MariaDBAccountSuiteTests instead? Or does it violate some assumption within the ginkgo execution model itself?

File: tests/functional/placementapi_controller_test.go
Comment (gibizer) [inline tests/functional/placementapi_controller_test.go:936]: Could you add a test that shows that when the account is changed the service config is regenerated with the new user and password and the hash in the statefulset's env is updated to trigger pod restarts?

File: controllers/placementapi_controller.go
Comment (stuggi) [reply controllers/placementapi_controller.go:359]: tls is used as soon the mariadbdatabase reflects it is hosted on a galera supporting tls and the service has a CA to validate it. we could add setting requiredTLS as a follow up. maybe after we switched our default to enable podlevel/internal tls per default?

Comment (gibizer) [reply controllers/placementapi_controller.go:359]: I'm OK about a follow up, just seemed strange that we merged #165  yesterday in placement, and here we explicitly disable it. So I wanted to make sure it is intentional.

File: controllers/placementapi_controller.go
Comment (zzzeek) [reply controllers/placementapi_controller.go:357]: Yes I noticed that the recent merge has consolidated DB functions to be ahead of configmaps in any case, which was one of the reasons ensureaccount was up there.   since ensureaccount isn't actually going away, for those controllers where ensuredb is before configmaps (Which I guess will be all of them now), might as well put ensure account into it.    it starts to look like the whole of the various ensureDB functions could just be one big function in mariadb-operator, but i dont know that we have time to do all that just yet.

File: tests/functional/placementapi_controller_test.go
Comment (zzzeek) [reply tests/functional/placementapi_controller_test.go]: not sure?  I was just doing this whatever way I could get the funcs passed along, what benefits does a struct offer?   the thing about populating structs is that it seems you can leave fields omitted which is what I dont want here, every field is required.

File: tests/functional/placementapi_controller_test.go
Comment (zzzeek) [reply tests/functional/placementapi_controller_test.go:936]: those assertions would be specific to placement here.   this would either be a test that's specific to the placement repo (and also the other repos that it becomes a part of), or maybe the mariadb harness here could offer a hook for additional suites based on an account change that leverages the existing fixture functions.

File: controllers/placementapi_controller.go
Comment (zzzeek) [reply controllers/placementapi_controller.go:359]: the ensuremariadbaccount is ultimately going to be a receiver of a MariaDBAccount that may be coming from openstack-operator, and the account is not associated with any database as of yet.   at this stage, it's not appropriate to set TLS, because nothing is yet known about what kind of database this account will be applied to.     if we are flipping a TLS flag on when we do the ensure MariaDBDatabase phase, then we would want to flip on "require TLS" on the MariaDBAccount as well, and that would be within the mariadbdatabase_funcs.   if that's the use we want?  that is, if DB supports TLS, then all accounts should *require* TLS.  is that a req?

Comment (stuggi) [reply controllers/placementapi_controller.go:359]: > I'm OK about a follow up, just seemed strange that we merged #165 yesterday in placement, and here we explicitly disable it. So I wanted to make sure it is intentional.

right, setting requireTLS == false is not that then no tls is used for db connection. it is just that this db user could use tls and non tls to connect. if tls is used is being controlled by the my.cnf content. 

>  that is, if DB supports TLS, then all accounts should _require_ TLS. is that a req?

I am not sure right now. I can discuss it in our tomorrows tls meeting and get back

File: tests/functional/placementapi_controller_test.go
Comment (zzzeek) [reply tests/functional/placementapi_controller_test.go]: OK struct has names.   So.... make one struct and send to the function, how about this?  give this the thumbs up and I can change to this, I added a declare/run/go/???? method which takes the helper and timeout/interval


```
	// Run MariaDBAccount suite tests.  these are pre-packaged ginkgo tests
	// that exercise standard account create / update patterns that should be
	// common to all controllers that ensure MariaDBAccount CRs.
	mariadb_suite := &mariadb_test.MariaDBAccountSuite{
		name:          "Placement",
		namespace:     names.PlacementAPIName.Namespace,
		databaseName:  placement.DatabaseName,
		finalizerName: "PlacementAPI",
		generateCR: func(accountName types.NamespacedName) {
			// ...
		},
		updateAccount: func(newAccountName types.NamespacedName) {

			Eventually(func(g Gomega) {
				placementapi := GetPlacementAPI(names.PlacementAPIName)
				placementapi.Spec.DatabaseAccount = newAccountName.Name
				g.Expect(th.K8sClient.Update(ctx, placementapi)).Should(Succeed())
			}, timeout, interval).Should(Succeed())

		},
		deleteCR: func() {
			th.DeleteInstance(GetPlacementAPI(names.PlacementAPIName))
		},
	}
        // DeclareGinkgo() ?  Run()?  
	mariadb_suite.DeclareSuite(
		mariadb, timeout, interval,
	)

```

File: controllers/placementapi_controller.go
Comment (gibizer) [reply controllers/placementapi_controller.go:359]: thanks for the explanation. I defer to @stuggi if this is OK as is.

File: controllers/placementapi_controller.go
Comment (gibizer) [reply controllers/placementapi_controller.go:357]: yes, @stuggi fixed the ordering of the steps to have db created before we generate config. I would do the ensureDB consolidation here locally, and maybe looking at a common mariadb-operator/api function later separatly

File: tests/functional/placementapi_controller_test.go
Comment (gibizer) [reply tests/functional/placementapi_controller_test.go]: on the optionality of the struct fields, that is a downside here yes, but you can add a runtime check for it.

Comment (gibizer) [reply tests/functional/placementapi_controller_test.go]: my soft preference is towards mariadb_suite.Run()

File: tests/functional/placementapi_controller_test.go
Comment (gibizer) [reply tests/functional/placementapi_controller_test.go:936]: I would go with the placement specific test case as that seems simpler for this than extra hooks in the harness.

Comment (zzzeek) [reply tests/functional/placementapi_controller_test.go:936]: I added additional helper suites in mariadb so here we can see it's doing just the placement specific parts of it.  this is done as an add-on as in other projects the URL / confighash can be broken up based on controller and whatnot.

**Review by gibizer — COMMENTED** [review]
looks good to me. We just need to land the mariadb dependency first

**Comment by zzzeek** [comment]
> looks good to me. We just need to land the mariadb dependency first

yah once mariadb lands I can re-crank through everyone and add mistral / telemetry / barbican also

File: tests/functional/placementapi_controller_test.go
Comment (zzzeek) [reply tests/functional/placementapi_controller_test.go]: I added additional "optional / run multiple times" suites to it as well so it now has `RunBasicSuite()` `RunURLAssertSuite(assertURL assertsURL)` `RunConfigHashSuite(getConfigHash getsConfigHash)` 

i think this makes it a lot easier to test everyone's ability to render expected URLs upfront/on change as well as being able to assert any number of confighashes that are expected to change

File: controllers/placementapi_controller.go
Comment (stuggi) [reply controllers/placementapi_controller.go:359]: we did not came up with a real preference if "require TLS" should be used as soon the DB supports TLS. since the services will use tls as soon the DB supports it and the service has the CA we could set it, but atm it won't hurt keep it available. I'd personally do it in a follow up, after we tested all the different scenarios, also with adoption in mind. adoption of tlse env is still in early phase.

Comment (zzzeek) [reply controllers/placementapi_controller.go:359]: yah i think this should be all followup stuff.  we're just trying to get all the levers available right now (like, every controller gets a distinct DB password, stuff like that :)  )

**Review by gibizer — APPROVED** [review]
I tested this and I was able to rotate the db password properly. Nice work!

**Pull request merged**
